MathLib/KelvinVector-impl.h

```diff
@@ -36,16 +36,23 @@ double Invariants<KelvinVectorSize>::J3(
     Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_v)
 {
     assert(std::abs(trace(deviatoric_v)) <=
-           1e-16 * std::abs(deviatoric_v.squaredNorm()));
+           5e-14 * diagonal(deviatoric_v).norm());
     return determinant(deviatoric_v);
 }
 
+template <int KelvinVectorSize>
+Eigen::Vector3d Invariants<KelvinVectorSize>::diagonal(
+    Eigen::Matrix<double, KelvinVectorSize, 1> const& v)
+{
+    return v.template topLeftCorner<3, 1>();
+}
+
 /// Trace of the corresponding tensor.
 template <int KelvinVectorSize>
 double Invariants<KelvinVectorSize>::trace(
     Eigen::Matrix<double, KelvinVectorSize, 1> const& v)
 {
-    return v.template topLeftCorner<3, 1>().sum();
+    return diagonal(v).sum();
```

Comment:
The former solution might have been more efficient, because `diagonal()` creates a temporary.

Reply:
Good point! I'll check this.

Reply:
Looks like there is no temporary. I made a small setup
```c++
    KelvinVectorType<3> const kv = KelvinVectorType<3>::Random();
    Invariants<6> const invariants;
    volatile double t = invariants.trace(kv);
```
and changed the implementation of trace btw. topLeftCorner and diagonal. There is no difference in the assembly code (in release mode).

I'd guess that because the diagonal method is inline (being a template), or return value optimization works, or both....


Anyway no difference.

If there are no further comments, I'll merge it tonight.